pkg/executor/docker/executor.go

```diff
@@ -95,277 +110,259 @@ func (e *Executor) Run(
 	ctx context.Context,
 	request *executor.RunCommandRequest,
 ) (*models.RunCommandResult, error) {
-	log.Ctx(ctx).Info().Msgf("running execution %s", request.ExecutionID)
-	ctx, cancel := context.WithCancel(ctx)
-	e.cancellers.Put(request.ExecutionID, cancel)
-	defer func() {
-		if cancelFn, found := e.cancellers.Get(request.ExecutionID); found {
-			e.cancellers.Delete(request.ExecutionID)
-			cancelFn()
-		}
-	}()
-
-	//nolint:ineffassign,staticcheck
-	ctx, span := system.NewSpan(ctx, system.GetTracer(), "pkg/executor/docker.Executor.Run")
-	defer span.End()
-	defer e.cleanupExecution(ctx, request.ExecutionID)
-
-	dockerArgs, err := dockermodels.DecodeSpec(request.EngineParams)
+	if err := e.Start(ctx, request); err != nil {
+		return nil, err
+	}
+	res, err := e.Wait(ctx, request.ExecutionID)
 	if err != nil {
 		return nil, err
 	}
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case out := <-res:
+		return out, nil
+	}
+}
 
-	e.activeFlags[request.ExecutionID] = make(chan struct{}, 1)
-
-	// the actual mounts we will give to the container
-	// these are paths for both input and output data
-	var mounts []mount.Mount
-	for _, input := range request.Inputs {
-		if input.Volume.Type == storage.StorageVolumeConnectorBind {
-			log.Ctx(ctx).Trace().Msgf("Input Volume: %+v %+v", input.InputSource, input.Volume)
-
-			mounts = append(mounts, mount.Mount{
-				Type:     mount.TypeBind,
-				ReadOnly: input.Volume.ReadOnly,
-				Source:   input.Volume.Source,
-				Target:   input.Volume.Target,
-			})
+// TODO need a better name as this starts an execution for the request, not the actual executor this method is on.
+// Launc, Exec, Dispatch?
+func (e *Executor) Start(ctx context.Context, request *executor.RunCommandRequest) error {
+	log.Ctx(ctx).Info().
+		Str("executionID", request.ExecutionID).
+		Str("jobID", request.JobID).
+		Msg("starting execution")
+
+	if handler, found := e.handlers.Get(request.ExecutionID); found {
+		if handler.active() {
+			// TODO we can make this method idempotent by not returning an error here if the execution is already active
+			return fmt.Errorf("execution (%s) already started", request.ExecutionID)
 		} else {
-			return executor.FailResult(fmt.Errorf("unknown storage volume type: %s", input.Volume.Type))
+			// TODO what should we do if an execution has already completed and we try to start it again? Just rerun it?
```

Comment:
Currently, it's not possible to go from any terminal state (cancelled, failed, complete) to a running state again.  

Except long-running jobs, if they end up in a failed state (due to a compute node restart) we should be able to restart the execution _without_ obtaining a new allocation from the requester node which may, or may not, be available.

Reply:
It might be too early to answer this, but would we expect a restarted execution (recovered execution?) to go through a this same code path? Or would there be a different method on the executor that would "recover" and start a previously failed execution?

For now do you think it makes sense to continue erroring here?

Reply:
I think this makes sense here, but I need to fix up the Startup so that for any long-running jobs it _only_ calls Run() (and not cancel/fail) and let's the executor deal with the idempotence.